fiftyone/core/dataset.py

```
@@ -1433,74 +1439,196 @@ def resolve_transformation(
-    def _resolve_transformation_direct(
-        self, sample, source_frame, target_frame, transforms
-    ):
-        """Directly resolve transform without chaining."""
-        # Check for sample-level transforms by matching type
+    def _get_direct_transformations(self, sample, transforms):
+        """Builds direct transforms keyed by ``(source_frame, target_frame)``."""
+        sample_transforms = {}
+        dataset_transforms = {}
+        world_shorthand = {}
+
+        def _add_transform(target, source_frame, target_frame, transform):
```

Ruff (0.15.9)

[warning] Missing return type annotation for private function `_add_transform`

Add return type annotation: `None`

(ANN202)